File: backend/app/services/storage.py
import hashlib
import os
import uuid
import aiofiles
from fastapi import UploadFile
from app.core.config import settings
from app.core.logging import logger


class StorageService:
    def __init__(self, base_dir: str = settings.STORAGE_DIR):
        self.base_dir = base_dir
        os.makedirs(self.base_dir, exist_ok=True)

    async def save_upload_file(
        self, workspace_id: str, upload_file: UploadFile
    ) -> tuple[str, str, int]:
        # Ensure workspace subdirectory exists
        ws_dir = os.path.join(self.base_dir, workspace_id)
        os.makedirs(ws_dir, exist_ok=True)

        # Generate unique filename to avoid collisions
        file_ext = os.path.splitext(upload_file.filename or "")[1].lower()
        if not file_ext:
            file_ext = ".pdf"

        unique_filename = f"{uuid.uuid4()}{file_ext}"
        target_path = os.path.join(ws_dir, unique_filename)

        sha256_hash = hashlib.sha256()
        file_size = 0

        async with aiofiles.open(target_path, "wb") as out_file:
            while chunk := await upload_file.read(1024 * 1024):
                file_size += len(chunk)
                sha256_hash.update(chunk)
                await out_file.write(chunk)

        checksum = sha256_hash.hexdigest()
        logger.info(f"Saved file {upload_file.filename} ({file_size} bytes) to {target_path}")
        return target_path, checksum, file_size

    def delete_file(self, storage_path: str) -> bool:
        try:
            if os.path.exists(storage_path):
                os.remove(storage_path)
                logger.info(f"Deleted file from storage: {storage_path}")
                return True
        except Exception as e:
            logger.warning(f"Failed to delete file {storage_path}: {str(e)}")
        return False


storage_service = StorageService()

File: backend_rust/src/services/storage.rs
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};
use tokio::fs;
use tokio::io::AsyncWriteExt;
use uuid::Uuid;
use crate::config::Settings;
use crate::error::{AppError, AppResult};

pub struct StorageService {
    base_dir: PathBuf,
    max_file_size: u64,
    allowed_extensions: Vec<String>,
}

impl StorageService {
    pub fn new(settings: &Settings) -> Self {
        let base_dir = settings.storage_dir.clone();
        let max_file_size = (settings.max_upload_size_mb as u64) * 1024 * 1024;
        let allowed_extensions = settings.allowed_extensions.clone();

        // Ensure base directory exists
        let base_dir_clone = base_dir.clone();
        tokio::spawn(async move {
            if let Err(e) = fs::create_dir_all(&base_dir_clone).await {
                tracing::error!("Failed to create storage directory: {}", e);
            }
        });

        Self {
            base_dir,
            max_file_size,
            allowed_extensions,
        }
    }

    pub async fn save_upload_file(
        &self,
        workspace_id: Uuid,
        mut file: tokio::fs::File,
        filename: &str,
    ) -> AppResult<(String, String, u64)> {
        // Validate file extension
        let ext = Path::new(filename)
            .extension()
            .and_then(|e| e.to_str())
            .map(|e| format!(".{}", e.to_lowercase()))
            .unwrap_or_else(|| ".pdf".to_string());

        if !self.allowed_extensions.iter().any(|allowed| *allowed == ext) {
            return Err(AppError::BadRequest(format!(
                "Unsupported file format '{}'. Allowed formats: {}",
                ext,
                self.allowed_extensions.join(", ")
            )));
        }

        // Create workspace subdirectory
        let ws_dir = self.base_dir.join(workspace_id.to_string());
        fs::create_dir_all(&ws_dir).await?;

        // Generate unique filename
        let unique_filename = format!("{}{}", Uuid::new_v4(), ext);
        let target_path = ws_dir.join(&unique_filename);

        // Calculate checksum and save file
        let mut hasher = Sha256::new();
        let mut file_size = 0u64;
        let mut buffer = vec![0u8; 1024 * 1024]; // 1MB buffer

        let mut out_file = fs::File::create(&target_path).await?;

        loop {
            let n = file.read(&mut buffer).await?;
            if n == 0 {
                break;
            }
            file_size += n as u64;

            if file_size > self.max_file_size {
                // Clean up partial file
                let _ = fs::remove_file(&target_path).await;
                return Err(AppError::PayloadTooLarge(format!(
                    "File size exceeds maximum allowed size of {}MB",
                    self.max_file_size / (1024 * 1024)
                )));
            }

            hasher.update(&buffer[..n]);
            out_file.write_all(&buffer[..n]).await?;
        }

        out_file.flush().await?;
        drop(out_file);

        let checksum = format!("{:x}", hasher.finalize());
        let storage_path = target_path.to_string_lossy().to_string();

        tracing::info!(
            "Saved file {} ({} bytes) to {}",
            filename,
            file_size,
            storage_path
        );

        Ok((storage_path, checksum, file_size))
    }

    pub fn delete_file(&self, storage_path: &str) -> bool {
        let path = Path::new(storage_path);
        if path.exists() {
            match std::fs::remove_file(path) {
                Ok(_) => {
                    tracing::info!("Deleted file from storage: {}", storage_path);
                    true
                }
                Err(e) => {
                    tracing::warn!("Failed to delete file {}: {}", storage_path, e);
                    false
                }
            }
        } else {
            false
        }
    }

    pub fn get_file_path(&self, storage_path: &str) -> Option<PathBuf> {
        let path = Path::new(storage_path);
        if path.exists() {
            Some(path.to_path_buf())
        } else {
            None
        }
    }
}